### PR TITLE
8308899: Introduce Classfile context and improve Classfile options

### PR DESCRIPTION
**Pull request by asotona**
Classfile context object and multi-state options have been discussed at https://mail.openjdk.org/pipermail/classfile-api-dev/2023-May/000321.html
This patch implements the proposed changes in Classfile API and fixes all affected code across JDK sources and tests.

Please review.

Thanks,
Adam

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8308899](https://bugs.openjdk.org/browse/JDK-8308899): Introduce Classfile context and improve Classfile options (**Sub-task** - P4)
 * [JDK-8306650](https://bugs.openjdk.org/browse/JDK-8306650): Improve control of stack maps generation in Classfile API (**Enhancement** - P4)
 * [JDK-8308646](https://bugs.openjdk.org/browse/JDK-8308646): Typo in ConstantValueAttribute (**Bug** - P5)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [5723e66a](https://git.openjdk.org/jdk/pull/14180/files/5723e66ac9204671d942a6b5f49310112e4fd2e9)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14180/head:pull/14180` \
`$ git checkout pull/14180`

Update a local copy of the PR: \
`$ git checkout pull/14180` \
`$ git pull https://git.openjdk.org/jdk.git pull/14180/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14180`

View PR using the GUI difftool: \
`$ git pr show -t 14180`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14180.diff">https://git.openjdk.org/jdk/pull/14180.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14180#issuecomment-1572201180)

### CODE REVIEW

**Review by liach — COMMENTED** [review]
A side comment about the API changes:
I think we should recommend users to store Classfile instances in static final variables if sharing is reasonable, like in Module and BindingSpecializer; this way, we can cache class hierarchy resolution information across builds.

File: src/java.base/share/classes/jdk/internal/classfile/Classfile.java
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/Classfile.java:206]: yes, several methods should move to ClassfileImpl, thanks.

File: src/java.base/share/classes/jdk/internal/classfile/components/ClassRemapper.java
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/components/ClassRemapper.java:102]: yes, I'll add the parameter, thanks.

File: src/java.base/share/classes/java/lang/Module.java
Comment (liach) [inline src/java.base/share/classes/java/lang/Module.java:1593]: This `cc` can be stored in a static final field instead.

File: src/java.base/share/classes/jdk/internal/classfile/Classfile.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/Classfile.java:55]: We can create a static final field in `ClassfileImpl` holding a default instance equivalent to that created by `new ClassfileImpl()` and have the `of()` factory return that instance instead.

File: src/java.base/share/classes/jdk/internal/classfile/impl/StackMapGenerator.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/StackMapGenerator.java:840]: This should set the Class hierarchy resolver, patch dead code, and filter dead labels options from the original context.

**Comment by liach** [comment]
In fact, I think we can probably mark Classfile as value-based. Then, whether it's created from a factory at every call or if it's cached (my first 2 review comments) no longer matters as much; the caching behavior will be handled by individual option instances (such as class hierarchy caching in the class hierarchy option's resolver)

File: src/java.base/share/classes/java/lang/Module.java
Comment (asotona) [reply src/java.base/share/classes/java/lang/Module.java:1593]: Yes, we can avoid repeated construction of default class hierarchy resolver here.

File: src/java.base/share/classes/jdk/internal/classfile/Classfile.java
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/Classfile.java:55]: Global static default context will hold global shared cached class hierarchy resolver and that we want to avoid.

File: src/java.base/share/classes/jdk/internal/classfile/impl/StackMapGenerator.java
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/impl/StackMapGenerator.java:840]: Here the other options are irrelevant. The error report wraps already generated incomplete code attribute into a fake class to be able to print details to the error.
I think even the stack map generation suppression is irrelevant here, will double check.

File: src/java.base/share/classes/java/lang/Module.java
Comment (asotona) [reply src/java.base/share/classes/java/lang/Module.java:1593]: In this situation the ClassHierarchyResolver is never used and have no caching function.
Adding static field with initializer does not make much sense here.

File: src/java.base/share/classes/jdk/internal/classfile/constantpool/ConstantPoolBuilder.java
Comment (briangoetz) [inline src/java.base/share/classes/jdk/internal/classfile/constantpool/ConstantPoolBuilder.java:82]: The old ctors here were (ClassModel) and (Options), where the ClassModel version had a trail of breadcrumbs back to options.  The new ctors are (ClassModel) and no-args, which is a slight departure.  What use did CPB make of options in the first place, if any?

**Review by briangoetz — COMMENTED** [review]
Overall this looks pretty clean.  I like how simply it flows through the implementation classes.  There are a number of methods and classes, especially in Classfile, which need some more spec.

File: src/java.base/share/classes/jdk/internal/classfile/Classfile.java
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/Classfile.java:52]: Fixed, thanks.

File: src/java.base/share/classes/jdk/internal/classfile/Classfile.java
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/Classfile.java:76]: Fixed, thanks.

File: src/java.base/share/classes/jdk/internal/classfile/Classfile.java
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/Classfile.java:79]: Fixed, thanks.

File: src/java.base/share/classes/jdk/internal/classfile/ClassHierarchyResolver.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/ClassHierarchyResolver.java:66]: Don't think we need to synchronize the cache if we have dedicated CHRs.

**Comment by briangoetz** [comment]
Agreed.  The need for sync came from the shared static cache.

**Comment by asotona** [comment]
Unfortunately thread-unsafe context makes sharing of it in tests executed in parallel a nightmare.
I can fix our Corpus tests and hope the race condition won't raise also somewhere else later.
However how to explain this limitation to users?

I suggest to make it always thread-safe (as the context is primary expected to be shared in multi-threaded environment) and users may make it faster in specific cases by providing non-synchronized map.

**Comment by briangoetz** [comment]
Most entities in the system are immutable and therefore freely sharable 
without additional coordination.  Some have lazily-inflated single-field 
caches that conform to the "benign race" (there is only one possible 
non-default value) criteria.

The main hole is the CHA resolution cache.  There are two ways to 
address this: protect the cache, or don't share the context.

Unfortunately now that ClassModel refers back to the context, there's a 
risk of unexpected context sharing.  Here are a few options:

  - Make the CHA cache thread-safe using, say, CHM;
  - Make the CHA cache unshared, by using a ThreadLocal (this brings the 
context back to being immutable)
  - Detect when implicit use of a context happens on a thread other than 
the creating thread (currently the only vector for this is 
Classfile::transform), and inflate a new context with empty cache in 
that case;
  - Always create a fresh context with an empty cache in 
Classfile::transform

**Comment by asotona** [comment]
>  Here are a few options: 
 - Make the CHA cache thread-safe using, say, CHM; 

This is current solution for the default system CH cache. 

We may also use a bit less strict custom semi-synchronization (just enough to be thread safe) instead of fully-synchronized `CHM::computeIfAbsent`, for example `CHM:get … (compute if absent without lock) ... CHM::put`.
CHM does not lock on get, so the only penalty would be synchronous put call. 

> - Make the CHA cache unshared, by using a ThreadLocal (this brings the context back to being immutable)

This is interesting option for small number of threads, however it makes big footprint for heavy-parallel systems.

> - Detect when implicit use of a context happens on a thread other than the creating thread (currently the only vector for this is Classfile::transform), and inflate a new context with empty cache in that case;

This would degrade the caching completely by triggering a new context for each transformed method in case there are two threads working synchronously.

> - Always create a fresh context with an empty cache in Classfile::transform

This would destroy the caching. Such cache would work only in scope of one method.

**Comment by briangoetz** [comment]
On 6/8/2023 12:09 PM, Adam Sotona wrote:
>
>     Here are a few options:
>      - Make the CHA cache thread-safe using, say, CHM;
>
> This is current solution for the default system CH cache.
>
> We may also use a bit less strict custom semi-synchronization (just 
> enough to be thread safe) instead of fully-synchronized 
> |CHM::computeIfAbsent|, for example |CHM:get … (compute if absent 
> without lock) ... CHM::put|.
> CHM does not lock on get, so the only penalty would be synchronous put 
> call.
>
>      - Make the CHA cache unshared, by using a ThreadLocal (this
>     brings the context back to being immutable)
>
> This is interesting option for small number of threads, however it 
> makes big footprint for heavy-parallel systems.
>

I have a hard time imagining heavy parallel use here; concurrency 
against the same cache would most likely come from accidental sharing.  
So this is not necessarily a problem.  Are you imagining differnet use 
cases?

**Comment by asotona** [comment]
> I have a hard time imagining heavy parallel use here; concurrency against the same cache would most likely come from accidental sharing.  So this is not necessarily a problem.  Are you imagining differnet use cases?

I think it is common practice for example in web services to pre-configure as much of the common parts as possible (and Classfile context logically falls into that category) and then register the service to the server, which use group of thread workers to handle individual requests.
With (non-reusable) virtual threads and structured concurrency on the table makes internal use of thread locals even less sense. I can imagine a new virtual thread launched (and finished) for each individual transformation. Our `CorpusTest` relying on Junit parallelization  is not so far from that scenario (I'm not sure if Junit already uses virtual threads or just a group of workers).

It is technically irrelevant if user will share the context directly or through a `ClassModel` (for example an instrumentor class used to transform all incoming code - this BTW an idea of faster `j.l.r.ProxyGenerator` I'm thinking about). I think we should be sure that `Classfile` context (and also expansion of `ClassModel`) does not fail in parallel environments.

**Comment by asotona** [comment]
I'll run our standard benchmarks to compare following scenarios:
- static synchronized single instance cache (current master)
- context-based non-synchronized caches  `HM::computeIfAbsent`
- context-based fully synchronized caches `CHM::computeIfAbsent`
- context-based thread-safe lazy synchronized caches `CHM::get ... compute ... CHM::put`

And let's see the performance impact.

**Comment by asotona** [comment]
I've performed a lot of benchmarking and the key result is that introduction of `Classfile` context cause a lot of regressions in our benchmarks.

Below are aggregated only relative results of benchmarks with significant regressions.
Current `master` with global static cache instance is a base for the relative regressions in the table.
- `ConcurrentHashMap::computeIfAbsent` represents the initial fully-synchronized cache implementation.
- `HashMap::computeIfAbsent` is the thread-unsafe option.
- `ConcurrentHashMap::get and put` is the lazy-synchronized solution
- **Always create a fresh context with an empty cache** is what Brian proposed as one of the options
- **Class loading with local cache only** is a step back to what ASM is using.


Interesting point is that **Always create a fresh context with an empty cache** works with similar performance as the others, except for `GenerateStackMaps.benchmark`, where it causes significant regression.

**Class loading with local cache only** assumes `ClassLoader` would do the caching for us. Even the class loading is initially slower and does not work for all use cases, it may be the best for the default CHR. And we could skip any `CHR::cached` in the API. 

Also interesting is that `ConcurrentHashMap::computeIfAbsent` cause 10% regression of **ParseOptions.transformNoStackmap** benchmark, where it should have no or minimal effect.

  | `ConcurrentHashMap::computeIfAbsent` | `HashMap::computeIfAbsent` | `ConcurrentHashMap::get and put` | Always create a fresh context with an empty cache | Class loading with local cache only
-- | -- | -- | -- | -- | --
AdHocAdapt.transform LIFT | -19% | -17% | -16% | -17% | -17%
AdHocAdapt.transform LIFT1 | -24% | -25% | -23% | -26% | -23%
AdHocAdapt.transform LIFT2 | -23% | -26% | -23% | -25% | -23%
AdaptInjectNoop.transform NOP_SHARED | -16% | -19% | -16% | -18% | -7%
AdaptNull.transform SHARED_3 | -18% | -21% | -19% | -17% | -18%
AdaptNull.transform SHARED_3_NO_DEBUG | -22% | -22% | -25% | -21% | -22%
GenerateStackMaps.benchmark |   |   |   | -27% |  
ParseOptions.transformNoDebug | -31% | -20% | -20% | -19% | -18%
ParseOptions.transformNoLineNumbers | -25% | -24% | -20% | -20% | -20%
ParseOptions.transformNoStackmap | -10% |   |   |   |

**Comment by briangoetz** [comment]
FWIW: I have found AdaptNull.SHARED_3 to be among the most useful 
benchmarks for finding unexpected regressions.

I think there's something fishy going on.  Given that the baseline 
(static shared cache) has to protect against concurrent access already, 
and that there should be little or no other significant difference 
between the static version and the context version aside from caching, I 
can't see how those differences explain these regressions.  Are we 
making an apples-to-apples comparison of what CHR algorithm we're using? 
Is the creation of the CHR instances is perturbing the results?

Let's try to take CHR operation out of the equation: runt both baseline 
and context versions using a CHR that does almost nothing (e.g., says 
LUB(a,b) is always Object) or which uses a precomputed static cache.  
I'm guessing this shows no regression, and if so, this will likely tell 
us that the _context_ refactor is not the problem, but attempting to 
unshare the CHR cache.  Then we can drill into that problem.

File: src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java:334]: On the baseline, if a stack map table attribute can be reused but this findAttribute doesn't find an attribute (implying an implicit stack map, potentially for simple code?), we will not perform a stack map generation and simply reuse a trivial stack map (null), which should be a correct algorithm.

File: src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java:349]: As a result (see previous comment), we will run into a redundant stack map generation when the original stack map table is trivial (absent). Maybe we want to consider user-configured selective stack map regeneration to replace this faulty logiic.

**Review by liach — COMMENTED** [review]
I have found the cause of the regression: there's redundant stack map generation in DirectCodeBuilder for methods with trivial stack map table (null).

**Comment by liach** [comment]
Since "not generating stack maps" really means to use an implicit stack map, I suggest renaming the `STACK_MAPS_NEVER` to `STACK_MAPS_IMPLICIT` to indicate such a confusing feature of StackMapTable. (In fact, the enum element names of StackMapsOption doesn't fit in with other options, they should be ALWAYS_STACK_MAPS, NEVER_STACK_MAPS, etc.)

File: src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java:334]: Yes, I caused this regression by "Revert of DirectCodeBuilder.needsStackMap pre-calculation" commit above.
The needStackMap pre-calculation on the other side caused regression in lambda performance, so one regression has been replaced with another.

File: src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java]: I've fixed the logic in the commit below.

**Comment by asotona** [comment]
All the performance regressions have been found and fixed.
Stackmaps generation logic is following:
- `STACK_MAPS_WHEN_REQUIRED` is the default option and performance-critical. It keeps original stack maps if code match and generate them only when required.
- `STACK_MAPS_ALWAYS` always (re)generate stack maps no matter if exist in the original
- `STACK_MAPS_NEVER` stack maps are not generated and dropped if already exist

Naming convention can be still adjusted.

Default `ClassHierarchyResolver` now use reflection from system class loader and it is not globally cached (so no dilemma about CHM vs HM). Any CHR is cached on method level if not already cached.

Benchmarks comparison of this PR vs master now shows even slight improvement:
```
Benchmark                                  (transform)    master  this PR Units
AdHocAdapt.transform                              LIFT     8.704    9.629 ops/s
AdHocAdapt.transform                             LIFT1     8.622    9.275 ops/s
AdHocAdapt.transform                             LIFT2     9.257    9.378 ops/s
AdaptInjectNoop.transform                   NOP_SHARED     4.709    4.978 ops/s
AdaptNull.transform                           SHARED_1    96.686   98.027 ops/s
AdaptNull.transform                           SHARED_2    48.820   49.257 ops/s
AdaptNull.transform                           SHARED_3     9.611    9.691 ops/s
AdaptNull.transform                  SHARED_3_NO_DEBUG    12.218   12.343 ops/s
AdaptNull.transform                         UNSHARED_3     3.614    3.603 ops/s
ParseOptions.transformNoDebug                      N/A    11.169   11.084 ops/s
ParseOptions.transformNoLineNumbers                N/A    10.590   10.523 ops/s
ParseOptions.transformNoStackmap                   N/A     6.220    9.646 ops/s
```

**Comment by asotona** [comment]
>StackMapsOption doesn't fit in with other options, they should be ALWAYS_STACK_MAPS, NEVER_STACK_MAPS, etc.)

What about `STACK_MAPS_WHEN_REQUIRED`, `GENERATE_STACK_MAPS` and `DROP_STACK_MAPS`?

File: src/java.base/share/classes/jdk/internal/classfile/Classfile.java
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/Classfile.java:61]: As the default CHR stays static instance and it is no more cached, we can have also static holder for the default Classfile instance now.

**Comment by asotona** [comment]
I think we've resolved controversy about global static CHR cache (by having no global static CHR cache ;) and turned all performance issues into minor improvements. A new `ClassfileBenchmark` has been added for more precise future continuous performance monitoring.

This PR is now waiting for review to unblock all follow-up conversion work.

Thank you,
Adam

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassfileImpl.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/ClassfileImpl.java:63]: Don't users of the default context still share a non-thread-safe-cached version of the default resolver? We might still need to convert DEFAULT_CONTEXT field to a static defaultContext() method that returns a `new ClassfileImpl(...);` (the attribute mapper option can be shared).

Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/ClassfileImpl.java:63]: Please disregard this comment: Now the distinction of cache of resolvers is done in `StackMapGenerator`, meaning each stack map generation task has its own cache. Didn't see it early on.

File: src/java.base/share/classes/jdk/internal/classfile/Classfile.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/Classfile.java:343]:
```suggestion
     *     this.build(model.thisClass(), ConstantPoolBuilder.of(model),
     *                     b -> b.transform(model, transform));
```

File: src/java.base/share/classes/jdk/internal/classfile/Classfile.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/Classfile.java:272]: The CP is now simply a data holder; it no longer holds the processing options. Same update is needed for the other few build methods taking a CP instance.
```suggestion
     * builder.
```

File: src/java.base/share/classes/jdk/internal/classfile/Classfile.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/Classfile.java:57]:
```suggestion
     * {@return a context with default options}
```

**Comment by liach** [comment]
@asotona I've sent a patch to your fork for a few API problems spotted a while ago and reported on the mailing list. Feel free to review my patch when you are back to work.

**Comment by liach** [comment]
Since you fixed the string format typos, can you fix this typo as well? https://bugs.openjdk.org/browse/JDK-8308646 and remember to include that issue with `/issue add JDK-8308646` if you do.

**Comment by liach** [comment]
I ran `make docs-classfile-api-javadoc` and the results look good in general. There are a few spec and API issues scattered across the whole API, which I think I will probably collect and list them in the mailing list later.

One thing of note is that the JDK itself has a `@sealedGraph` taglet to render the sealed class hierarchy of any class. We might want to add that tag to all our sealed public classes or interfaces. But it requires a somewhat more complicated build process (including generating dot files with graphviz and rendering them to svg) and you might tweak the `SetupExecute` to a full-on `SetupApiDocsGeneration`.

**Review by mcimadamore — APPROVED** [review]
The API changes look great. I like the instance usages of ClassFile, and how CF is now used in a centralized fashion to double down as a cache/option store (and entry point for transformation). Seems like a step in a good direction, and make the API feel a little more cohesive.

A small subjective comment - I'm not too fond of using "context" as a variable/accessor name for all the places where we use a classfile. E.g. if classfile is the context, perhaps these variable/accessor should just be named "classfile" ?

**Comment by mcimadamore** [comment]
> The API changes look great. I like the instance usages of ClassFile, and how CF is now used in a centralized fashion to double down as a cache/option store (and entry point for transformation). Seems like a step in a good direction, and make the API feel a little more cohesive.
> 
> A small subjective comment - I'm not too fond of using "context" as a variable/accessor name for all the places where we use a classfile. E.g. if classfile is the context, perhaps these variable/accessor should just be named "classfile" ?

Also, another observation: the number of updates snippets in javadoc seems rather low for something that touches the main entry point by which client interacts with classfiles. This seems a sign of perhaps not having many end to end examples in the API javadoc?

**Comment by asotona** [comment]
@mcimadamore thank you for review!
Description of `Classfile` instances as "context" is intentional compromise between discoverability (as originally proposed `Classfile.Context` didn't work very well) and clarity of the function (as `Classfile` instances may be simply confused with `ClassModel` instances).

**Comment by asotona** [comment]
> I ran `make docs-classfile-api-javadoc` 

Thanks for pointing this out. The makefile patches accidentally fell into this PR.

> One thing of note is that the JDK itself has a `@sealedGraph` taglet to render the sealed class hierarchy of any class. 

I'll look at adding the tags, thanks.

**Comment by liach** [comment]
Congrats, now I can update many of my patches that depend on this API update.

**Pull request closed without merging**
